### PR TITLE
Use `module_function` for `Homebrew` module.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Use `module_function` rather than `extend self` to convert instance methods to class methods.

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
Can you explain a bit why https://github.com/Homebrew/homebrew-test-bot/pull/2 is needed?

**Comment by reitermarkus** [comment]
This is needed because without `extend self`, the methods inside `module Homebrew` blocks will not be converted to class methods globally, but rather on a per-block basis.

**Comment by MikeMcQuaid** [comment]
@reitermarkus I see. I think we need to avoid breaking the previous API because we support it for third-party commands and can't update them globally.

**Comment by reitermarkus** [comment]
We could put 

```ruby
module Homebrew
  extend self
end
```

in `compat/global.rb`.

**Comment by MikeMcQuaid** [comment]
@reitermarkus That works for me. Is there any way to detect usage of the old functions and print a warning on them somehow? We don't need to do that eventually but trying to figure out how we'd deprecate that eventually.

**Comment by reitermarkus** [comment]
Took me quite a while to figure this out (because of course I had a typo in `module Hombrew` while testing), but I found a way.

**Comment by penman** [comment]
fwiw `module_function` will (IIRC) prevent you having any methods that are `private` or `protected` to the module, whereas this works with `extend self`.

**Comment by reitermarkus** [comment]
@penman, `private_class_method`: https://github.com/Homebrew/brew/pull/1145/files#diff-3003aac9196017d495a416e89622da87R58

For `protected` you'll have to use `class << self` though.

**Comment by penman** [comment]
TIL @reitermarkus :D

**Comment by MikeMcQuaid** [comment]
A style failure here but it looks like the bot is behaving at least: https://bot.brew.sh/job/Brew%20Pull%20Requests/2103/version=any_sierra/testReport/junit/brew-test-bot/sierra/style/

**Comment by reitermarkus** [comment]
The style failure seems to be a bug in RuboCop. Will open an issue.

**Comment by reitermarkus** [comment]
If someone wants to test this, uncomment the `odeprecated` line and try running the `test-bot`, while [the PR over there](https://github.com/Homebrew/homebrew-test-bot/issues/2) isn't merged.

File: Library/Homebrew/cmd/link.rb [hunk 1]
```diff
@@ -71,11 +73,10 @@ def link
...
   end
 
-  private
-
   def keg_only?(rack)
     Formulary.from_rack(rack).keg_only?
   rescue FormulaUnavailableError, TapFormulaAmbiguityError, TapFormulaWithOldnameAmbiguityError
     false
   end
+  private_class_method :keg_only?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/link.rb]: I think for the `cmd`s we can remove the `private_class_method`s as we don't really care if they are private or not.

File: Library/Homebrew/compat/global.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    if instance_methods.include?(method)
+      # odeprecated "#{self}##{method}", "'module_function' or 'def self.#{method}' to convert it to a class method"
+      instance_method(method).bind(self).call(*args, &block)
+    else
+      super
+    end
+  end
+
+  def respond_to_missing?(method, include_private = false)
+    if method_defined?(method)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/compat/global.rb]:
```
return true if method_defined?(method)
super(method, include_private)
```

File: Library/Homebrew/compat/global.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+module Homebrew
+  module_function
+
+  def method_missing(method, *args, &block)
+    if instance_methods.include?(method)
+      # odeprecated "#{self}##{method}", "'module_function' or 'def self.#{method}' to convert it to a class method"
+      instance_method(method).bind(self).call(*args, &block)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/compat/global.rb]: Maybe `return` this and skip the `else`?

File: Library/Homebrew/global.rb [hunk 4]
```diff
@@ -39,14 +39,25 @@
 require "tap_constants"
 
 module Homebrew
-  include FileUtils
-  extend self
+  module_function
 
-  attr_accessor :failed
```
Comment (MikeMcQuaid) [inline Library/Homebrew/global.rb]: There's no equivalent?

Comment (reitermarkus) [inline Library/Homebrew/global.rb]: I didn't check, but why would you keep both?

`failed == failed?`

Comment (MikeMcQuaid) [inline Library/Homebrew/global.rb]: Sorry, not this specifically but just being able to use `attr_accessor`

Comment (reitermarkus) [inline Library/Homebrew/global.rb]:
```
class << self
  attr_accessor :something
end
```

Comment (MikeMcQuaid) [inline Library/Homebrew/global.rb]: Reasons to not do that here?

Comment (reitermarkus) [inline Library/Homebrew/global.rb]: No, just didn't check before. :+1:

**Comment by MikeMcQuaid** [comment]
I'm 🆒 with 🚢 ing this when you are @reitermarkus when you're confident in it.

**Pull request merged**